libraries/AP_Beacon/AP_Beacon_Nooploop.cpp

```diff
@@ -0,0 +1,261 @@
+#define NOOPLOOP_SF0_SZ 128 // setting_frame0 packet size
+
+#define NOOPLOOP_HEADER                     0x55    // message header
+#define NOOPLOOP_FUNCTION_MARK_NODE_FRAME2  4
+#define NOOPLOOP_NODE_FRAME2_FRAMELEN_MAX   4096    // frames should be less than 4k bytes
+#define NOOPLOOP_NODE_FRAME2_SYSTIME        6       // start of 4 bytes holding system time in ms
+#define NOOPLOOP_NODE_FRAME2_PRECISION_X    10      // start of 1 byte holding precision in m*100 in x axis
+#define NOOPLOOP_NODE_FRAME2_PRECISION_Y    11      // start of 1 byte holding precision in m*100 in y axis
+#define NOOPLOOP_NODE_FRAME2_PRECISION_Z    12      // start of 1 byte holding precision in m*100 in y axis
+#define NOOPLOOP_NODE_FRAME2_POSX           13      // start of 3 bytes holding x position in m*1000
+#define NOOPLOOP_NODE_FRAME2_POSY           16      // start of 3 bytes holding y position in m*1000
+#define NOOPLOOP_NODE_FRAME2_POSZ           19      // start of 3 bytes holding z position in m*1000
+#define NOOPLOOP_NODE_FRAME2_VALID_NODES    118
+#define NOOPLOOP_NODE_FRAME2_NODE_BLOCK     119
+
+#define NOOPLOOP_HEADER2                    0x54    // message header
+#define NOOPLOOP_FUNCTION_MARK_SETTING_FRAME0  0
+#define NOOPLOOP_SETTING_FRAME0_A0          37
+
+
+extern const AP_HAL::HAL& hal;
+
+AP_Beacon_Nooploop::AP_Beacon_Nooploop(AP_Beacon &frontend, AP_SerialManager &serial_manager) :
+    AP_Beacon_Backend(frontend)
+{
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Beacon, 0);
+    if (_uart != nullptr) {
+        _uart->begin(serial_manager.find_baudrate(AP_SerialManager::SerialProtocol_Beacon, 0));
+    }
+}
+
+// return true if sensor is basically healthy (we are receiving data)
+bool AP_Beacon_Nooploop::healthy()
+{
+    // healthy if we have parsed a message within the past 300ms
+    return ((AP_HAL::millis() - _last_update_ms) < AP_BEACON_TIMEOUT_MS);
+}
+
+// update the state of the sensor
+void AP_Beacon_Nooploop::update(void)
+{
+    // return immediately if not serial port
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check uart for any incoming messages
+    uint32_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        uint8_t b = _uart->read();
+        MsgType type = parse_byte(b);
+        if (type == MsgType::NODE_FRAME2) {
+            if (_anchor_pos_avail) {
+                parse_node_frame2();
+            } else if (AP_HAL::millis() - _last_request_setting_ms > 2000) {
+                _last_request_setting_ms = AP_HAL::millis();
+                request_setting();
+            }
+        } else if (type == MsgType::SETTING_FRAME0) {
+            parse_setting_frame0();
+        } 
+    }
+}
+
+void AP_Beacon_Nooploop::request_setting()
+{
+    //send setting_frame0 to tag, tag will fill anchor position and ack
+    _uart->write((uint8_t)0x54);
+    _uart->write((uint8_t)0);
+    _uart->write((uint8_t)1);
+    for (int i = 0; i < 124; i++) {
```

Comment:
sorry to be annoying but this "int" should be a uint16_t probably.

Reply:
Thank you. Fixed